src/Grpc.AspNetCore.Server/Internal/ServerCallDeadlineManager.cs

```diff
@@ -0,0 +1,263 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class ServerCallDeadlineManager : IAsyncDisposable
+    {
+        // Max System.Threading.Timer due time
+        private const long DefaultMaxTimerDueTime = uint.MaxValue - 1;
+
+        // Avoid allocating delegates
+        private static readonly TimerCallback DeadlineExceededDelegate = DeadlineExceededCallback;
+        private static readonly TimerCallback DeadlineExceededLongDelegate = DeadlineExceededLongCallback;
+
+        private readonly Timer _longDeadlineTimer;
+        private readonly ISystemClock _systemClock;
+        private readonly HttpContextServerCallContext _serverCallContext;
+
+        private CancellationTokenSource? _deadlineCts;
+        private CancellationTokenRegistration _requestAbortedRegistration;
+        private TaskCompletionSource<object?>? _deadlineExceededCompleteTcs;
+
+        public DateTime Deadline { get; }
+        public bool IsCallComplete { get; private set; }
+        public bool IsDeadlineExceededStarted => _deadlineExceededCompleteTcs != null;
+
+        public CancellationToken CancellationToken
```

Comment:
What's actually using this token now? Just app code?

Reply:
Yes. `ServerCallContext.CancellationToken` is exposed to devs.